tree/dataframe/src/RCsvDS.cxx

```diff
@@ -187,10 +202,45 @@ std::vector<void *> RCsvDS::GetColumnReadersImpl(std::string_view colName, const
 
 void RCsvDS::InferColTypes(std::vector<std::string> &columns)
 {
-   auto i = 0U;
-   for (auto &col : columns) {
-      InferType(col, i);
-      ++i;
+   for (auto const &col : fColTypes) {
+      if (!HasColumn(col.first)) {
+         std::string msg = "There is no column with name \"" + col.first + "\".";
+         if (!fReadHeaders) {
+            msg += "\nSince the input csv file does not contain headers, valid column names";
+            msg += " are [\"Col0\", ..., \"Col" + std::to_string(columns.size() - 1) + "\"].";
+         }
+         throw std::runtime_error(msg);
+      }
+      if (std::string("bdls").find(col.second) == std::string::npos) {
+         std::string msg = "Type alias '" + std::string(1, col.second) + "' is not supported.\n";
+         msg += "Supported type aliases are 'b' for boolean, 'd' for double, 'l' for Long64_t, 's' for std::string.";
+         throw std::runtime_error(msg);
+      }
+   }
+
+   const auto second_line = fCsvFile->GetFilePos();
+   for (auto i = 0u; i < columns.size(); ++i) {
+      if (fColTypes.find(fHeaders[i]) == fColTypes.end()) { // type is not yet specified
+         if (columns[i] == fFillValue) {
+            // read 10 lines until a non-empty cell on this column is found, so that type is determined
+            // if 10 lines are read, then type is automatically assumed `double`
+            for (auto row = 0u; row < 10u; ++row) {
+               std::string line;
+               if (fCsvFile->Readln(line)) {
+                  const auto temp_columns = ParseColumns(line);
+                  if (temp_columns[i] != fFillValue) {
+                     columns[i] = temp_columns[i];
+                     break;
+                  }
+               } else { // whole column was empty
+                  break;
+               }
+            }
+            // reset the reading from the second line, because the first line is already loaded in `columns`
+            fCsvFile->Seek(second_line);
+         }
+         InferType(columns[i], i);
```

Comment:
this code has some deep nesting. it might be more readable if we refactor it to be less nested (e.g. this version has 3 levels instead of 6):

```suggestion
   for (auto i = 0u; i < columns.size(); ++i) {
      if (fColTypes.find(fHeaders[i]) != fColTypes.end())
         continue; // type was manually specified, nothing to do
      auto extraRowsRead = 0u;
      while (columns[i] == fFillValue && extraRowsRead < 10) {
         std::string line;
         if (!fCsvFile->Readln(line))
            break; // EOF
         ++extraRowsRead;
         const auto temp_columns = ParseColumns(line);
         if (temp_columns[i] != fFillValue)
            columns[i] = temp_columns[i];
      }
      fCsvFile->Seek(second_line);

      if (columns[i] == fFillValue)
         columns[i] = 'd'; // could not find a non-empty value, default to double
      else
          InferType(columns[i], i);
```

Reply:
This looks great. I am applying it with the minor difference of using a `for` loop instead of `while` - saving 2 extra lines :) Plus `++extraRowsRead;` looks like being on somehow random position now (in the middle of the inner body)

Reply:
I even made it better!